src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1219,6 +1220,8 @@ RouterInsertTaskList(Query *query, DeferredErrorMessage **planningError)
 
 	Assert(query->commandType == CMD_INSERT);
 
+	NormalizeInsertTargetList(query);
```

Comment:
If we're doing shard pruning in the planner, then we'll evaluate the expressions in `values_lists` in the executor in `ExecuteMasterEvaluableFunctions`, but during deparsing we'll use the unevaluated expressions from the `rowValuesLists`.

If we're doing shard pruning in the executor, then we're already post-`ExecuteMasterEvaluableFunctions` at this point. `ExecuteMasterEvaluableFunctions` will incorrectly evaluate the target list entries once, and then move the resulting constants to the `values_lists`.

Reply:
Yeah I realized that right after I pushed this up. I still am not 100% certain what you're describing is entirely accurate (largely due to the use of a lot of in-place evaluation in this area) but it would certainly be much clearer to pull out the normalize method I added (assuming it is changed to be correct) and call it before doing `ExecuteMasterEvaluableFunctions`.

Reply:
You'll also want to be careful that during planning time we work with a rewritten query (with constant expressions evaluated), while during execution time we work with the original query (with expressions evaluated by our own logic).

Reply:
I think there are several places here the distinction between `originalQuery` and the planned one are blurred and it's leading to some confusion, i.e. I think I removed some code thinking there were deep references from the `Task` values lists into the actual values of a `Query`, such that evaluating the latter (in place) would update the former. If that ever held, it certainly doesn't anymore.

Basically I think either (a) the bit that I added to evaluate `values_lists` expressions in `ExecuteMasterEvaluableFunctions` should be removed, or (b) `values_lists` should be replaced with `NIL` before that is called. In addition, we should just evaluate the values lists for each task separately.

So:

  * Fix normalization such that `Var` order in the target list matches 1—1 the values in `values_list` (either by injecting the non-`Var` nodes where they belong into `values_lists` entries or by putting our reconstructed `Var`-based `TLE` on the end of the `TLE`)
  * Ensure master evaluation does not evaluate the `values_lists` in the query
  * Instead directly evaluate the `values_lists` that we need to, as we are executing each row

I suppose this improvement could be left until after release…

Reply:
This is why I settled on a hack.  :)